packages/ember-application/lib/system/application.js

```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
   },
 
   /**
-    Calling initialize manually is not supported.
-
-    Please see Ember.Application#advanceReadiness and
-    Ember.Application#deferReadiness.
+    Initialize the application and return a promise that resolves with the `Ember.Application`
+    object when the boot process is complete.
 
-    @private
-    @deprecated
-    @method initialize
-   **/
-  initialize() {
-    deprecate('Calling initialize manually is not supported. Please see Ember.Application#advanceReadiness and Ember.Application#deferReadiness');
-  },
+    Run any application initializers and run the application load hook. These hooks may
+    choose to defer readiness. For example, an authentication hook might want to defer
+    readiness until the auth token has been retrieved.
 
-  /**
-    Initialize the application. This happens automatically.
+    By default, this method is called automatically on "DOM ready"; however, if autoboot
+    is disabled, you would need to call this method yourself. Calling this method before
+    "DOM ready" is not recommended.
 
-    Run any initializers and run the application load hook. These hooks may
-    choose to defer readiness. For example, an authentication hook might want
-    to defer readiness until the auth token has been retrieved.
-
-    @private
-    @method domReady
+    @public
+    @method boot
+    @return {Promise<Ember.Application,Error>}
   */
-  domReady() {
-    if (this.isDestroyed) { return; }
+  boot() {
+    if (this._bootPromise) { return this._bootPromise; }
 
-    this.boot();
+    try {
+      this._bootSync();
+    } catch(_) {
+      // Ignore th error: in the asynchronous boot path, the error is already reflected
```

Comment:
This is :cry:

Basically that's part of the reason why _bootSync uses `defer` instead of wrapping it in a promise. Code paths that treats boot as synchronous (now calls `_bootSync`) *might* expect exceptions, so `_bootSync` as an API/method needs to throw. On the other hand, the async API wants to reify this as a promise rejection.

So `_bootSync` and `boot` works like this:

```javascript
boot() {
  // Async!
  try {
    _bootSync();
  } catch(_) {
    // Nope!
  }

  return promiseCreatedByBootSync;
}

_bootSync() {
    var defer = new Defer();

    try {
      doStuff();
      defer.resolve(this);
    } catch(e) {
      defer.reject(e);
      throw e; // for existing consumers
    }
}
```